src/common/FastRational.h

```diff
@@ -739,19 +804,50 @@ inline void multiplication(FastRational& dst, const FastRational& a, const FastR
 }
 
 inline void division(FastRational& dst, const FastRational& a, const FastRational& b) {
+    if (b.wordPartValid() && b.num == 1 && b.den == 1) {
+        dst = a;
+        return;
+    }
+    if (a.wordPartValid() && a.num == 0) {
+        dst = 0;
+        return;
+    }
     if (a.wordPartValid() && b.wordPartValid()) {
+        if (a.num == b.num && a.den == b.den) {
+            dst.num = 1;
+            dst.den = 1;
+            dst.setWordPartValid();
+            dst.kill_mpq();
+            return;
+        }
         uword common1 = gcd(absVal(a.num), absVal(b.num));
         uword common2 = gcd(a.den, b.den);
         word zn;
         uword zd;
-        CHECK_WORD(zn, (lword(a.num)/common1) * (b.den/common2));
+        CHECK_WORD(zn, ulword(absVal(a.num)/common1) * (b.den/common2));
         CHECK_UWORD(zd, ulword(absVal(b.num)/common1) * (a.den/common2));
+
+        // Note: dst and a or b might be the same FastRational.
+        bool b_num_lt_0 = b.num < 0;
+        bool a_num_ge_0 = a.num >= 0;
+        bool b_num_gt_0 = b.num > 0;
+        bool a_num_le_0 = a.num <= 0;
+
+        if ((b_num_lt_0 && a_num_ge_0) || (b_num_gt_0 && a_num_le_0)) { //CHECK_WORD(zn, (lword)(-absVal(zn)));
+            lword tmp = -(lword)(absVal(zn));
+            if (tmp < WORD_MIN || tmp > WORD_MAX) {
+                goto overflow;
+            }
+            zn = tmp;
+        }
+        else if ((b_num_gt_0 && a_num_ge_0) || (b_num_lt_0 && a_num_le_0)) CHECK_WORD(zn, (lword)(absVal(zn)));
```

Comment:
Also here (`absVal`)

Reply:
I refactored this based on the comment.  Hopefully nothing broke.